### ISSUE TITLE
Investigate Pagination Support

### ISSUE THREAD

**Issue by ffazzini** [derjust/spring-data-dynamodb#55]
Does the project support pagination without loading the whole data in memory first?

Example Request Params usage for the Pageable object:
?size=2&page=0

It seems to me that it is loading everything in memory before filtering the results because this code is invoked at some point:
https://github.com/derjust/spring-data-dynamodb/blob/master/src/main/java/org/socialsignin/spring/data/dynamodb/repository/query/AbstractDynamoDBQuery.java#L158

Is there a way to achieve this? Am I missing something? 
Thanks in advance,
Francesco

**Comment by derjust** [comment]
Yes it does. This is provided by the AWS SDK on its own (check out their `List` implementation`).
Longer answer to come but in short: Don't use any method calls that need the 'whole' list - like `size()` or `get()`.
As long as you access the list only iterative, it does paginate the result set

**Comment by ffazzini** [comment]
Thanks that's good. Is there a way to explicitly specify the ExclusiveStartKey and limit to achieve the pagination mechanism? Or should I use aws sdk directly for that?

**Comment by derjust** [comment]
It depends a little bit where you want to go:

* If you just want to iterate through the resultset and have a paged UI (web page with paginator) you may just use the Spring-Data `Pageable` and the rest is done for you under the hood
* If you want to iterative process the data, you even can go with the `List` version and let the SDK do the heavy lifting. Just don't call `length`, `get` and only `iterator().*`
* If you have to jump into the result set back you may want to use your own `CustomRepositoryImpl` and have access to the AWS SDK classes there and do all the details (But I always would argue that this is just a special case of #1 and with a proper range key & `findBy` that should work already)

**Comment by derjust** [comment]
More details: https://aws.amazon.com/blogs/developer/understanding-auto-paginated-scan-with-dynamodbmapper/
